finetuner/tuner/callback/early_stopping.py

```diff
@@ -0,0 +1,110 @@
+from typing import TYPE_CHECKING
+
+import numpy as np
+from jina.logging.logger import JinaLogger
+
+from finetuner.helper import get_framework
+from .base import BaseCallback
+
+
+if TYPE_CHECKING:
+    from ..base import BaseTuner
+
+
+class EarlyStopping(BaseCallback):
+    """
+    Callback to stop training when a monitored metric has stopped improving.
+    A `model.fit()` training loop will check at end of every epoch whether
+    the monitered metric is no longer improving.
+    """
+
+    def __init__(
+        self,
+        monitor: str = 'val_loss',
+        mode: str = 'auto',
+        patience: int = 2,
+    ):
+        """
+        :param monitor: if `monitor='loss'` best bodel saved will be according
+            to the training loss, if `monitor='val_loss'` best model saved will be
+            according to the validation loss
+        :param mode: one of {'auto', 'min', 'max'}. the
+            decision to overwrite the current save file is made based on either
+            the maximization or the minimization of the monitored quantity.
+            For `val_acc`, this should be `max`, for `val_loss` this should be
+            `min`, etc. In `auto` mode, the mode is set to `max` if the quantities
+            monitored are 'acc' or start with 'fmeasure' and are set to `min` for
+            the rest of the quantities.
+        :param patience: integer, the number of epochs after which the training is
+            stopped if there is no improvement
+        """
+        self._logger = JinaLogger(self.__class__.__name__)
+        self._monitor = monitor
+        self._mode = mode
+        self._patience = patience
+        self._train_losses = []
+        self._valid_losses = []
+        self._wait = 0
+
+        if mode not in ['auto', 'min', 'max']:
+            self._logger.warning(
+                'ModelCheckpoint mode %s is unknown, ' 'fallback to auto mode.', mode
+            )
+            mode = 'auto'
+
+        if mode == 'min':
+            self._monitor_op = np.less
+            self._best = np.Inf
+        elif mode == 'max':
+            self._monitor_op = np.greater
+            self._best = -np.Inf
+        else:
+            if 'acc' in self._monitor:  # to adjust other metrics are added
+                self._monitor_op = np.greater
+                self._best = -np.Inf
+            else:
+                self._monitor_op = np.less
+                self._best = np.Inf
+
+    def on_epoch_end(self, tuner: 'BaseTuner'):
```

Comment:
I'm implementing the methods of the abstract class BaseCallback so I don't think I can specify the function type hint and return at this level

Reply:
I don't think there is anything forbidding us to document functions. But you are right. :) If you do not except any other params than the base class, you can leave the type hints. 
But it would be nice, if you explained in a few words what your implementation of the function does.